libsolidity/analysis/FunctionCallGraph.h

```diff
@@ -0,0 +1,115 @@
+ *  - CreationDispatch: Represents the creation dispatch function
+ *  - ExternalDispatch: Represents the runtime dispatch for all external functions
+ *  - InternalDispatch: Represents the runtime dispatch for internal function pointers and complex expressions
+ *
+ *  Nodes are a variant of either the enum SpecialNode or an ASTNode pointer.
+ *  ASTNodes are usually inherited from CallableDeclarations
+ *  (FunctionDefinition, ModifierDefinition, EventDefinition) but for functions
+ *  without declaration it is directly the FunctionCall AST node.
+ *
+ *  Functions that are not called right away as well as functions without
+ *  declarations have an edge to the dispatch node. */
+class FunctionCallGraphBuilder: private ASTConstVisitor
+{
+public:
+	enum class SpecialNode { Unset, CreationRoot, CreationDispatch, InternalDispatch, ExternalDispatch };
+
+	typedef std::variant<ASTNode const*, SpecialNode> Node;
+
+	struct CompareByID
+	{
+		using is_transparent = void;
+
+		bool operator()(Node const& _lhs, Node const& _rhs) const
+		{
+			if (_lhs.index() != _rhs.index())
+				return _lhs.index() < _rhs.index();
+
+			if (std::holds_alternative<SpecialNode>(_lhs))
+				return std::get<SpecialNode>(_lhs) < std::get<SpecialNode>(_rhs);
+			return std::get<ASTNode const*>(_lhs)->id() < std::get<ASTNode const*>(_rhs)->id();
+		}
+		bool operator()(Node const& _lhs, int64_t _rhs) const
+		{
+			solAssert(!std::holds_alternative<SpecialNode>(_lhs), "");
+
+			return std::get<ASTNode const*>(_lhs)->id() < _rhs;
+		}
+		bool operator()(int64_t _lhs, Node const& _rhs) const
+		{
+			solAssert(!std::holds_alternative<SpecialNode>(_rhs), "");
+
+			return _lhs < std::get<ASTNode const*>(_rhs)->id();
+		}
+	};
+
+	struct ContractCallGraph
+	{
+		ContractCallGraph(ContractDefinition const& _contract) :contract(_contract) {}
+
+		/// Contract for which this is the graph
+		ContractDefinition const& contract;
+
+		std::map<Node, std::set<ASTNode const*, ASTNode::CompareByID>, CompareByID> edges;
+
+		/// Set of contracts created
+		std::set<ContractDefinition const*, ASTNode::CompareByID> createdContracts;
+	};
+
+	std::shared_ptr<ContractCallGraph> const create(ContractDefinition const& _contract);
+
+private:
+	bool visit(Identifier const& _identifier) override;
+	bool visit(NewExpression const& _newExpression) override;
+	void endVisit(MemberAccess const& _memberAccess) override;
+	void endVisit(FunctionCall const& _functionCall) override;
+
+	void visitCallable(CallableDeclaration const* _callable);
+	void visitConstructor(
+		ContractDefinition const& _contract,
+		std::vector<ContractDefinition const*>::const_iterator _start,
+		std::vector<ContractDefinition const*>::const_iterator _end
+	);
+
+	bool add(Node _caller, ASTNode const* _callee);
+	void processFunction(CallableDeclaration const& _callable, ExpressionAnnotation const& _annotation);
+
+	ContractDefinition const* m_contract = nullptr;
+	Node m_currentNode = SpecialNode::Unset;
```

Comment:
I think this should be `std::optional<Node> m_currentNode` instead of adding an `Unset` into the enum.